### PR TITLE
Switch to an easier pooling model for pooled collections in the IDE.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This follows the pattern we already have for ArrayBuilder and makes it easier to use these pooled collectoins without having to remember to free them (which we occasionally forget to do).

### CODE REVIEW

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 1]
```diff
@@ -3512,8 +3510,8 @@ private bool TryMapParameter((SyntaxNode? Node, int Ordinal) parameterKey, IRead
             //   the closure tree of the previous version and then map 
             //   closure scopes in the new version to the previous ones, keeping empty closures around.
 
-            var oldLocalCapturesBySyntax = PooledDictionary<SyntaxNode, int>.GetInstance();
-            var oldParameterCapturesByLambdaAndOrdinal = PooledDictionary<(SyntaxNode? Node, int Ordinal), int>.GetInstance();
+            using var _1 = PooledDictionary<SyntaxNode, int>.GetInstance(out var oldLocalCapturesBySyntax);
+            using var _2 = PooledDictionary<(SyntaxNode? Node, int Ordinal), int>.GetInstance(out var oldParameterCapturesByLambdaAndOrdinal);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: this was a bug that we didn't free this.

File: src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpDeclareAsNullableCodeFixProvider.cs [hunk 2]
```diff
@@ -65,15 +65,14 @@ private string GetEquivalenceKey(SyntaxNode node)
...
-            var alreadyHandled = PooledHashSet<TypeSyntax>.GetInstance();
+            using var _ = PooledHashSet<TypeSyntax>.GetInstance(out var alreadyHandled);
 
             foreach (var diagnostic in diagnostics)
             {
                 var node = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
                 MakeDeclarationNullable(editor, node, alreadyHandled);
             }
 
-            alreadyHandled.Free();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeFixes/Nullable/CSharpDeclareAsNullableCodeFixProvider.cs]: these are generally bugs as we might not reach here if we return early (i.e. if cancellationToken triggers).

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 3]
```diff
@@ -407,204 +407,194 @@ protected sealed override async Task FixAllAsync(Document document, ImmutableArr
...
-            var nodesToRemove = PooledHashSet<SyntaxNode>.GetInstance();
-            var nodesToAdd = PooledHashSet<(TLocalDeclarationStatementSyntax declarationStatement, SyntaxNode node)>.GetInstance();
+            using var _1 = PooledDictionary<SyntaxNode, SyntaxNode>.GetInstance(out var nodeReplacementMap);
+            using var _2 = PooledHashSet<SyntaxNode>.GetInstance(out var nodesToRemove);
+            using var _3 = PooledHashSet<(TLocalDeclarationStatementSyntax declarationStatement, SyntaxNode node)>.GetInstance(out var nodesToAdd);
             // Indicates if the node's trivia was processed.
-            var processedNodes = PooledHashSet<SyntaxNode>.GetInstance();
-            var candidateDeclarationStatementsForRemoval = PooledHashSet<TLocalDeclarationStatementSyntax>.GetInstance();
+            using var _4 = PooledHashSet<SyntaxNode>.GetInstance(out var processedNodes);
+            using var _5 = PooledHashSet<TLocalDeclarationStatementSyntax>.GetInstance(out var candidateDeclarationStatementsForRemoval);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: missed case.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledDictionary.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Microsoft.CodeAnalysis.PooledObjects
+{
+    internal sealed partial class PooledDictionary<K, V> : IPooled
+    {
+        public static PooledDisposer<PooledDictionary<K, V>> GetInstance(out PooledDictionary<K, V> instance)
+        {
+            instance = GetInstance();
+            return new PooledDisposer<PooledDictionary<K, V>>(instance);
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/ObjectPools/PooledDictionary.cs]: IDE only helper following pattern we already use.

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 5]
```diff
@@ -961,7 +961,7 @@ void Emit()
             out ImmutableArray<(Guid ThreadId, ActiveInstructionId OldInstructionId, LinePositionSpan NewSpan)> activeStatementsInUpdatedMethods,
             out ImmutableArray<(ActiveMethodId Method, NonRemappableRegion Region)> nonRemappableRegions)
         {
-            var changedNonRemappableSpans = PooledDictionary<(int MethodToken, int MethodVersion, LinePositionSpan BaseSpan), LinePositionSpan>.GetInstance();
+            using var _1 = PooledDictionary<(int MethodToken, int MethodVersion, LinePositionSpan BaseSpan), LinePositionSpan>.GetInstance(out var changedNonRemappableSpans);
```
Comment (jmarolf) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs]: I think you just want `_` here

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 6]
```diff
@@ -1030,7 +1030,7 @@ void AddNonRemappableRegion(LinePositionSpan oldSpan, LinePositionSpan newSpan,
             activeStatementsInUpdatedMethods = activeStatementsInUpdatedMethodsBuilder.ToImmutableAndFree();
 
             // Gather all active method instances contained in this project/module that are not up-to-date:
-            var unremappedActiveMethods = PooledHashSet<ActiveMethodId>.GetInstance();
+            using var _2 = PooledHashSet<ActiveMethodId>.GetInstance(out var unremappedActiveMethods);
```
Comment (jmarolf) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs]: I think you just want `_` here

File: src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs [hunk 7]
```diff
@@ -226,8 +226,8 @@ private static (SyntaxTriviaList sameLine, SyntaxTriviaList endOfLine) SplitTrai
             var localVariables = ArrayBuilder<ISymbol>.GetInstance();
 
             // Map a symbol to an index into the statementsFromDeclarationToEnd array.
-            var variableDeclarationIndex = PooledDictionary<ISymbol, int>.GetInstance();
-            var lastVariableUsageIndex = PooledDictionary<ISymbol, int>.GetInstance();
+            using var _1 = PooledDictionary<ISymbol, int>.GetInstance(out var variableDeclarationIndex);
+            using var _2 = PooledDictionary<ISymbol, int>.GetInstance(out var lastVariableUsageIndex);
```
Comment (jmarolf) [inline src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs]: I think you just want `_` here

File: src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -226,8 +226,8 @@ private static (SyntaxTriviaList sameLine, SyntaxTriviaList endOfLine) SplitTrai
             var localVariables = ArrayBuilder<ISymbol>.GetInstance();
 
             // Map a symbol to an index into the statementsFromDeclarationToEnd array.
-            var variableDeclarationIndex = PooledDictionary<ISymbol, int>.GetInstance();
-            var lastVariableUsageIndex = PooledDictionary<ISymbol, int>.GetInstance();
+            using var _1 = PooledDictionary<ISymbol, int>.GetInstance(out var variableDeclarationIndex);
```
Comment (jmarolf) [inline src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs]: I think you just want `_` here

File: src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs (hunk 7 above)
Comment (jmarolf) [reply src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs]: oh nevermind, I see that you need to do this for the using

**Comment by tmat** [comment]
>             var newCapturesToClosureScopes = ArrayBuilder<SyntaxNode?>.GetInstance(newCaptures.Length, null);

This one is also not freed on when we return in the middle of the method.

---
Refers to: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3141 in 0077844. [](commit_id = 007784487207f0847a80346b256eb9834a86c68d, deletion_comment = False)

**Comment by tmat** [comment]
>             var oldCapturesToClosureScopes = ArrayBuilder<SyntaxNode?>.GetInstance(oldCaptures.Length, null);

The same.

---
Refers to: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3145 in 0077844. [](commit_id = 007784487207f0847a80346b256eb9834a86c68d, deletion_comment = False)

**Comment by tmat** [comment]
>             var reverseCapturesMap = ArrayBuilder<int>.GetInstance(newCaptures.Length, 0);

This one as well.

---
Refers to: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3138 in 0077844. [](commit_id = 007784487207f0847a80346b256eb9834a86c68d, deletion_comment = False)

**Comment by CyrusNajmabadi** [comment]
> This one as well.

Yes.  I'm not touching ArrayBuilder as well.   That can come in a future PR.  IT's a lot more to cover/fix as ArrayBuilder is used *extensively* in the IDE code.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**